Source/Core/Core/PowerPC/Expression.cpp

```diff
@@ -0,0 +1,269 @@
+        if (index >= 0 && index <= 31 && end == name.data() + name.length())
+        {
+          bind.type = name[0] == 'r' ? VarBindingType::GPR : VarBindingType::FPR;
+          bind.index = index;
+        }
+      }
+      else if (name == "lr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_LR;
+      }
+      else if (name == "ctr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_CTR;
+      }
+      else if (name == "pc")
+      {
+        bind.type = VarBindingType::PCtr;
+      }
+    }
+
+    m_binds.emplace_back(bind);
+  }
+}
+
+std::optional<Expression> Expression::TryParse(std::string_view text)
+{
+  ExprVarListPointer vars{new expr_var_list{}};
+  ExprPointer ex{expr_create(text.data(), text.length(), vars.get(), g_expr_funcs.data())};
+  if (!ex)
+    return std::nullopt;
+
+  return Expression{text, std::move(ex), std::move(vars)};
+}
+
+double Expression::Evaluate() const
+{
+  SynchronizeBindings(SynchronizeDirection::From);
+
+  double result = expr_eval(m_expr.get());
+
+  SynchronizeBindings(SynchronizeDirection::To);
+
+  Reporting(result);
+
+  return result;
+}
+
+void Expression::SynchronizeBindings(SynchronizeDirection dir) const
+{
+  auto bind = m_binds.begin();
+  for (auto* v = m_vars->head; v != nullptr; v = v->next, ++bind)
+  {
+    switch (bind->type)
+    {
+    case VarBindingType::Zero:
+      if (dir == SynchronizeDirection::From)
+        v->value = 0;
+      break;
+    case VarBindingType::GPR:
+      if (dir == SynchronizeDirection::From)
+        v->value = static_cast<double>(GPR(bind->index));
+      else
+        GPR(bind->index) = static_cast<u32>(static_cast<s64>(v->value));
+      break;
+    case VarBindingType::FPR:
+      if (dir == SynchronizeDirection::From)
+        v->value = rPS(bind->index).PS0AsDouble();
+      else
+        rPS(bind->index).SetPS0(v->value);
+      break;
+    case VarBindingType::SPR:
+      if (dir == SynchronizeDirection::From)
+        v->value = static_cast<double>(rSPR(bind->index));
+      else
+        rSPR(bind->index) = static_cast<u32>(static_cast<s64>(v->value));
+      break;
+    case VarBindingType::PCtr:
+      if (dir == SynchronizeDirection::From)
+        v->value = static_cast<double>(PC);
+      break;
+    }
+  }
+}
+
+void Expression::Reporting(const double result) const
+{
+  bool is_nan = std::isnan(result);
+  std::string message;
+
+  for (auto* v = m_vars->head; v != nullptr; v = v->next)
+  {
+    if (isnan(v->value))
+      is_nan = true;
+
+    fmt::format_to(std::back_inserter(message), "  {}: {}", v->name, v->value);
+  }
+
+  if (is_nan)
+  {
+    message.append("\nBreakpoint condition encountered a NaN");
+    Core::DisplayMessage("Breakpoint condition has encountered a NaN.", 2000);
+  }
+
+  if (result != 0.0 || is_nan)
+    NOTICE_LOG_FMT(MEMMAP, "Breakpoint condition returned: {}. Vars: {}", result, message);
```

Comment:
I think `" {}={}"`using equal and removing one space would make it easier to read, the log string would then become `Vars:{}` *(without the space)*.